RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,440 @@
+        else {
+            // For `RowMajor` ordering
+            for (int64_t i=0; i < d; i++) {
+                if (selected_rows[i] != top) {
+                    blas::swap(cols, &A[cols * selected_rows[i]], 1, &A[cols * top], 1);
+                }
+                top++;
+            }
+        }
+    }
+
+    template<typename T, SignedInteger sint_t>
+    void permuteColsToLeft(
+                          blas::Layout layout,
+                          int64_t rows,
+                          int64_t cols,
+                          sint_t* selected_cols,
+                          int64_t d, // size of `selectedRows`
+                          T* A
+                          ) {
+        int64_t left = 0;  // Keeps track of the topmost unselected column
+
+        if(layout == blas::Layout::ColMajor) {
+            for (int64_t i=0; i < d; i++) {
+                if (selected_cols[i] != left) {
+                    // Use BLAS::swap to swap entire columns at once
+                    // Swapping col 'selected' with col 'top'
+                    blas::swap(rows, &A[rows * selected_cols[i]], 1, &A[rows * left], 1);
+                }
+                left++;
+            }
+        }
+        else {
+            // For `RowMajor` ordering
+            for (int64_t i=0; i < d; i++) {
+                if (selected_cols[i] != left) {
+                    blas::swap(rows, &A[selected_cols[i]], cols, &A[left], cols);
+                }
+                left++;
+            }
+        }
+    }
+
+    template <typename T>
+    void fht_left_col_major(T *buf, int log_n, int num_rows, int num_cols) {
+        int n = 1 << log_n;
+
+        // Apply FHT to each column independently
+        for (int col = 0; col < num_cols; ++col) {
+            // Pointer to the beginning of the current column in the Column-Major order
+            double* col_buf = buf + col * num_rows;
+
+            // Apply the original FHT on this column
+            for (int i = 0; i < log_n; ++i) {
+                int s1 = 1 << i;
+                int s2 = s1 << 1;
+                for (int j = 0; j < n; j += s2) {
+                    for (int k = 0; k < s1; ++k) {
+                        // For implicitly padding the input we just have to make sure
+                        // we replace all out-of-bounds accesses with zeros
+                        bool b1 = j + k < num_rows;
+                        bool b2 = j + k + s1 < num_rows;
+                        double u = b1 ? col_buf[j + k] : 0;
+                        double v = b2 ? col_buf[j + k + s1] : 0;
+                        if(b1 && b2) {
+                            col_buf[j + k] = u + v;
+                            col_buf[j + k + s1] = u - v;
+                        }
+                        else if(!b2 && b1) {
+                            col_buf[j + k] = u + v;
+                        }
+                        else if(!b2 && !b1)
+                            continue;
+                    }
```

Comment:
@aryamanjeendgar, sorry, jumped the gun. Please replace the ``double`` types here with ``T``. You'll need to do this in the other functions as well.

Reply:
Thank you for pointing this out! Corrected it!